### PR TITLE
feat: save Explorer queries to localStorage

### PR DESCRIPTION
**Pull request by cseas**
## Description
Initial PR for implementation of https://github.com/hypertrace/hypertrace-ui/issues/1551

### Testing
- Save Query button only appears when saved-queries feature is enabled.
- Save Query button is disabled if not filters are selected.
- Notification pops up when a query is saved successfully.
- Saved query is available in localStorage.
- Edge case: Saving a query works when no earlier saved preferences are available.

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### Documentation
Not sure if I should update the docs for this feature since we're merging to `rzp_main`.

### Screenshot

Screenshot 2022-05-11 at 12 35 37 PM

### Errors

```ts
this.notificationService.createSuccessToast('Query Saved Successfully!');
```

This line of code throws the following error in the console, although the notification pops up correctly on the screen.

Screenshot 2022-05-06 at 8 30 54 AM

I've created an Issue to discuss this:
https://github.com/hypertrace/hypertrace-ui/issues/1592

### CODE REVIEW

**Review by akashmane1598 — APPROVED** [review]
LGTM

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 1]
```diff
@@ -168,6 +191,23 @@ export class ExplorerComponent {
...
+  }
+
+  public onClickSaveQuery(): void {
+    this.preferenceService.get('savedQueries', []).subscribe(queries => {
+      this.savedQueries = queries as SavedQuery[];
+    });
+
+    const currentScope =
+      this.filters[0]?.metadata.scope === SPAN_SCOPE ? ScopeQueryParam.Spans : ScopeQueryParam.EndpointTraces;
+    const currentFilters = this.filters.map(filter => filter.urlString);
```
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts]: Rather can getting the current scope this way, can we read the scope of the explorer by making `ExplorerGeneratedDashboardContext` a `BehaviourSubject` and getting its current value?

That will make this relatively easier to read and understand IMO. Happy to discuss around it.

File: projects/components/src/filtering/filter/filter-attribute.ts [hunk 2]
```diff
@@ -7,4 +7,5 @@ export interface FilterAttribute {
   type: FilterAttributeType;
   onlySupportsAggregation?: boolean;
   onlySupportsGrouping?: boolean;
+  scope?: string;
```
Comment (jaywalker21) [inline projects/components/src/filtering/filter/filter-attribute.ts]: Do we need this `scope` entry here given we have created a type called `SavedQuery` which contains both `scope` and `filter`?

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 1 above)
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts]: I think this block and the one that follows should be within the subscribe part? 
Also, are we gaining anywhere by maintaining `this.savedQueries` given that we are always getting `savedQueries` from `preferenceService` and setting it again. I think it can be a simple `const` within the scope of the handler.

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 3]
```diff
@@ -168,6 +191,23 @@ export class ExplorerComponent {
...
+
+    this.savedQueries.push({ scope: currentScope, filters: currentFilters });
```
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts]: Sorry for checking this and telling about this now, but found something called `buildNavParamsWithFilters` in `ExplorerService`. I think that would be much simpler to use when we want to apply query. Currently, it saves the whole filter with all the metadata rather than just the `urlString`. Can you have a look at it once? I think it might be easier to save the whole filter in the storage and then call upon this method with `scope` and respective filter array.

Or we can add another method to the same service which takes an array of `filterUrlString` and `scope` and navigates to explorer.

Open for discussion around the same.

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 4]
```diff
@@ -168,6 +191,23 @@ export class ExplorerComponent {
...
+    this.preferenceService.set('savedQueries', this.savedQueries);
```
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts]: If the `set` is an observable action, we should probably do the `toast` in the subscribe part? That way in case something goes wrong setting it, we wouldn't show the notification.

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 5]
```diff
@@ -339,3 +379,8 @@ const enum ExplorerQueryParam {
   GroupLimit = 'limit',
   Series = 'series'
 }
+
+export interface SavedQuery {
+  scope: ScopeQueryParam;
+  filters: string[];
```
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts]: I think `filters` here is not the ideal word, maybe something like `filterUrlStrings`? Given we are only storing the `urlString` for filters.

**Review by jaywalker21 — COMMENTED** [review]
@cseas Have put some comments, whenever you can have a look.

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 1 above)
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: There's a `currentContext$` observable that already has the value of the scope. But we still need to do the ternary condition anyway because the context has different values than what we need exactly for the URL.

Values we get from currentContext$: `"API_TRACE"` or `"SPAN"`
Values we want: `"endpoint-traces"` or `"spans"`

Should I add a subscription to `currentContext$` in the constructor and put this ternary condition there?

File: projects/components/src/filtering/filter/filter-attribute.ts (hunk 2 above)
Comment (cseas) [reply projects/components/src/filtering/filter/filter-attribute.ts]: The `scope` entry here is for the `metadata` property which is of type `FilterAttribute`:
https://github.com/razorpay/hypertrace-ui/pull/32/files#diff-f19fe761c23b87fda9dc84db21cb18ead475913118bdff3f4f2e7af39990754bR205

Added the entry to avoid the typescript error saying: "scope doesn't exist on metadata"

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 1 above)
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: Just noticed there was already a method for this: `getQueryParamFromContext()`
Replaced the logic with that method here ✅

Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: Moved the subscription to the constructor instead of onClick so we subscribe only once. And added a const instead to the onClick handler ✅

File: projects/components/src/filtering/filter/filter-attribute.ts (hunk 2 above)
Comment (cseas) [reply projects/components/src/filtering/filter/filter-attribute.ts]: I replaced `metadata.scope` with `currentContext`. This type modification is not needed anymore. Removed ✅

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 4 above)
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: There's no subscribe() available on the set() method:
Screenshot 2022-05-13 at 2 39 58 PM

Should I add a try-catch here instead?

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 3 above)
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: Need to check after using the `buildNavParamsWithFilters` method and seeing if it works well.

Will address this in the next PR: https://github.com/razorpay/hypertrace-ui/pull/35
Have left a review comment for the same. ✅

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 5 above)
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: This field seems to map 1-to-1 to the `filter` parameter required by the `navigate` method in `NavigationService`. I'm using it like this in the follow-up PR:

Screenshot 2022-05-16 at 11 56 22 AM

But I agree, the naming shouldn't raise any confusion. Renaming it to `filterUrlStrings`.

Usage will change to `filter: query.filterUrlStrings` in the above screenshot.

Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: Done ✅

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 4 above)
Comment (jaywalker21) [reply projects/observability/src/pages/explorer/explorer.component.ts]: I guess we can let it be in that case. A `try....catch` block might not be needed, at least for in browser storage.
When we revise to have storage over DB using some network call, then we might need it, though it maybe within `PreferenceService`.

File: projects/observability/src/pages/explorer/explorer.component.ts [hunk 6]
```diff
@@ -118,11 +137,14 @@ export class ExplorerComponent {
   private static readonly VISUALIZATION_EXPANDED_PREFERENCE: string = 'explorer.visualizationExpanded';
   private static readonly RESULTS_EXPANDED_PREFERENCE: string = 'explorer.resultsExpanded';
   private readonly explorerDashboardBuilder: ExplorerDashboardBuilder;
+  private savedQueries: SavedQuery[] = [];
+  private currentContext: ExplorerGeneratedDashboardContext = ObservabilityTraceType.Api;
```
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts:141]: I think we can make a do without having this `currentContext` variable here.
As to how, we can change the `contextChangeSubject` (not `currentContext$`)  from `Subject` to `BehaviorSubject`.
`BehaviorSubject` are `Subject` which directly give us their current value without being subscribed to. Given both of them are referring to type `ExplorerGeneratedDashboardContext`, we can do `contextChangeSubject.getValue()`. That will also facilitate setting the initial value :)

Comment (jaywalker21) [reply projects/observability/src/pages/explorer/explorer.component.ts:141]: Also, it's more of a nitpick comment, so you can ignore it if you want.

**Review by jaywalker21 — APPROVED** [review]
One comment, rest all looks good.

File: projects/observability/src/pages/explorer/explorer.component.ts (hunk 6 above)
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts:141]: I tried changing the Subject to BehaviourSubject by changing

```ts
  private readonly contextChangeSubject: Subject<ExplorerGeneratedDashboardContext> = new Subject();
```

to

```ts
  private readonly contextChangeSubject: BehaviorSubject<ExplorerGeneratedDashboardContext> = new BehaviorSubject(
    ObservabilityTraceType.Api
  );
```

But TypeScript is going crazy with errors:

Screenshot 2022-05-16 at 2 02 40 PM

Am I doing something wrong?

Comment (jaywalker21) [reply projects/observability/src/pages/explorer/explorer.component.ts:141]: @cseas 

Ohh, there are two ways to handle this, that would resolve everything. You can bake it in future PRs if it makes sense.

```typescript
private readonly contextChangeSubject: BehaviorSubject<ExplorerGeneratedDashboardContext> = new BehaviorSubject(ObservabilityTraceType.Api as ExplorerGeneratedDashboardContext);
```

OR

```typescript
private readonly contextChangeSubject: BehaviorSubject<ExplorerGeneratedDashboardContext> = new BehaviorSubject<ExplorerGeneratedDashboardContext>(ObservabilityTraceType.Api);
```

Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts:141]: That worked. Thanks!

Change added to the next PR in the below commit:

https://github.com/razorpay/hypertrace-ui/pull/35/commits/ae6eea6bb3c993482bc86f64e98bbc2ea6bd665a

But changing `Subject` to `BehaviorSubject` started failing an existing test, couldn't figure out why so I reverted the change.

Here's the failing test:

https://github.com/razorpay/hypertrace-ui/runs/6450485969?check_suite_focus=true#step:4:1235
